File: src/Optimisation/Stochastic/BayesOpt.c

#include "bayesopt/bayesopt.h"

double inline_c_Optimisation_Stochastic_BayesOpt_0_c5fd47da4e55a10d4b997d4962b73a460a31de25(double val_inline_c_0) {
return ( val_inline_c_0 );
}


int inline_c_Optimisation_Stochastic_BayesOpt_1_e919766f7e72fc7f01b52086539aeac9e27e9078(double * vrx_inline_c_0, int dim_c_inline_c_1, double (* funIO_inline_c_2)(unsigned n, const double * x, double * gradien, void * fd), double * clv_inline_c_3, double * cuv_inline_c_4) {

            bopt_params bp;	// structure definition ??
            bp = initialize_parameters_to_default();
            set_learning(&bp, "L_MCMC");
            double *r = vrx_inline_c_0;
            bayes_optimization(
                dim_c_inline_c_1,
                funIO_inline_c_2,
                (void *) 0,	// we don't really need this argument, as we pass our a
                clv_inline_c_3,
                cuv_inline_c_4,
                r+1, r, bp);
        
}

File: src/Optimisation/Stochastic/BayesOpt.hs
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE QuasiQuotes #-}

module Optimisation.Stochastic.BayesOpt (
    Callback,
    bayesOptim
) where

import           Foreign.C.Types
import           Foreign.ForeignPtr (newForeignPtr_)
import           Foreign.Ptr
import           Data.Coerce (coerce)
import qualified Data.Vector.Storable as V
import qualified Data.Vector.Storable.Mutable as VM
import           Data.Monoid
import qualified Language.C.Inline as C

-- Haskell callback
type Callback a = a -> V.Vector Double -> IO Double

-- C callback:
type CCB a = CUInt -> Ptr CDouble -> Ptr CDouble -> Ptr a -> IO CDouble

-- | Possible parameters for optimisation algorithm
data Params = Kernel | Mean | Criteria | Surrogate | LogFile | LoadFile | SaveFile
            | Learning | Score

-- inline-c contexts needed for function pointers & vectors
C.context (C.baseCtx <> C.vecCtx <> C.funCtx)

C.include "bayesopt/bayesopt.h"

bayesOptim
    :: Callback a
    -> a
    -> V.Vector Double
    -> V.Vector Double
    -> IO (Double, V.Vector Double)
bayesOptim f a lv uv = do
    let dim   = V.length lv
        dim_c = fromIntegral dim
        clv   = coerce lv :: V.Vector CDouble
        cuv   = coerce uv :: V.Vector CDouble
    -- This is the C callback function, where we make some parameter
    -- conversions and then call the user callback (a Haskell function):
    let funIO :: CCB a
        funIO _dim xv _gv _a = do
            -- We use only the x vector, all other params we either don't need or have
            -- them already in Haskell
            ptr' <- newForeignPtr_ xv
            vec  <- V.freeze $ VM.unsafeFromForeignPtr0 ptr' dim
            val  <- coerce $ f a $ coerce vec
            [C.exp| double { $(double val) } |]
    -- A vector where we collect the minimum and the corresponding parameters:
    vrx <- VM.new (dim+1) :: IO (VM.IOVector CDouble)
    -- Here we call - after some further preparations - the C optimisation function
    -- It will take a looong time
    ret <- [C.block| int {
            bopt_params bp;
            bp = initialize_parameters_to_default();
            // Ad-hoc parameter settings - will have to expose some functions for this
            bp.n_iterations   = 25;
            bp.noise          = 0.5;
            bp.n_iter_relearn = 5;
            bp.n_init_samples = 5;
            set_learning(&bp, "L_MCMC");
            double *r = $vec-ptr:(double *vrx);
            bayes_optimization(
                $(int dim_c),
                $fun:(double (* funIO) (unsigned int n, const double *x, double *grad, void *fd)),
                (void *) 0,  // we don't really need this argument, as we pass our a
                $vec-ptr:(double *clv),
                $vec-ptr:(double *cuv),
                r+1, r, bp);
        } |]
    rx <- V.freeze vrx
    return (coerce $ V.head rx, coerce $ V.tail rx)
